File: lib-cli/tmp/server-ts/src/producer/events/disconnected.ts
import {
    Identification,
    Filter,
    Producer,
    Context,
    Protocol,
} from "../implementation/events";
import { Output } from "../implementation/events/disconnected";

export function emit(
    consumer: Identification,
    filter: Filter,
    context: Context,
    producer: Producer
): Promise<Output> {
	const user = context.removeUser(consumer.uuid());
	if (user instanceof Error) {
		return Promise.reject(user);
	}
	return Promise.resolve(
		new Output()
			.broadcast(filter.except(consumer.uuid()))
			.EventsMessage(
				new Protocol.Events.Message({
					user: user.name,
					uuid: consumer.uuid(),
					message: `User ${user.name} has been left chat`,
					timestamp: BigInt(Date.now()),
				})
			)
			.broadcast(filter.except(consumer.uuid()))
			.EventsUserDisconnected(
				new Protocol.Events.UserDisconnected({
					username: user.name,
					uuid: consumer.uuid(),
				})
			)
	);
}

File: examples/producer/rust/src/producer/events/disconnected.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use clibri::server;
use uuid::Uuid;

type BroadcastEventsMessage = Option<(Vec<Uuid>, protocol::Events::Message)>;
type BroadcastEventsUserDisconnected = (Vec<Uuid>, protocol::Events::UserDisconnected);

#[allow(unused_variables)]
pub async fn emit<E: server::Error, C: server::Control<E>>(
    scope: &mut Scope<'_, E, C>,
) -> Result<(BroadcastEventsMessage, BroadcastEventsUserDisconnected), String> {
    let uuid = scope.identification.uuid();
    if let Some(user) = scope.context.remove_user(uuid).await {
        let msg = scope
            .context
            .add_message(&user.name, format!("User {} has been left", user.name))?;
        Ok((
            Some((scope.filter.except(&uuid), msg.clone())),
            (
                scope.filter.except(&uuid),
                protocol::Events::UserDisconnected {
                    username: msg.user,
                    uuid: uuid.to_string(),
                },
            ),
        ))
    } else {
        Err(format!("User {} doesn't exist", uuid))
    }
}
